### ISSUE TITLE
Sentry NLog Target discards the events in case of large RPM

### ISSUE THREAD

**Issue by Shtannikov** [getsentry/sentry-dotnet#384]
Please mark the type framework used:

- [x] ASP.NET MVC
- [ ] ASP.NET Web API (OWIN)
- [x] ASP.NET Core
- [ ] WPF
- [ ] WinForms
- [ ] Xamarin
- [ ] Other:

Please mark the type of the runtime used:

- [x] .NET Framework
- [ ] Mono
- [x] .NET Core

* Version: 4.8 & 3.1

Please mark the NuGet packages used:

- [ ] Sentry
- [ ] Sentry.Serilog
- [x] Sentry.NLog
- [ ] Sentry.Log4Net
- [ ] Sentry.Extensions.Logging
- [ ] Sentry.AspNetCore

* Version: 2.0.3

These are the specs we have:
- up to 5000 RPM from our service 
- ~30 milliseconds response time for event logging in our On Premise Sentry

Under these conditions more than 50% of events don't reach Sentry.

We configured maxQueueItems for 10.000 and our load test showed that the Sentry Client queue had been overfulled 4 minutes after the launch.
We don't want to lose events but we are also afraid that configuring it for more than 10.000 may lead to performance issues of our service (e.g. OOM exceptions).

We ran load tests with Sentry Client being created for each logging request (hence per event) and they resulted success: nothing had been lost and latency was acceptable.
That's why we want to be able to turn off the reusable feature of Sentry client in Target. We are aware of that Breadcrumbs may suffer but we're okay with that.

We discussed the feature earlier (https://github.com/getsentry/sentry-dotnet/issues/370) but decided it wasn't reasonable enough. But now I feel like the queue problem could make it so.
Thoughts?

**Comment by snakefoot** [comment]
One of the great things about having a queue, is the ability to perform batching (sending multiple events in one operation). 

Making 10000 http-requests to send 10000 log-events, is slower than making 100 http-requests with a batch of 100 log-events.

Of course the problem with batching is that a lot of messages are lost when a batch is lost.

**Comment by Shtannikov** [comment]
I don't see batching feature in Sentry Target. And my Nginx metrics say that logs one event per request.
So can you please reveal what you mean?

**Comment by snakefoot** [comment]
I reveal that it could be a good idea to implement batching.


Sent from my Sony Xperia


---- Evgeny Shtannikov wrote ----


I don't see batching feature in Sentry Target. And my Nginx metrics say that logs one event per request.
So can you please reveal what you mean?

**Comment by bruno-garcia** [comment]
We could batch events of have multiple in-flight in parallel (as opposed to the current implementation that sends one at a time). But before we go ahead with that I'd like to understand what is the realistic load you expect of exceptions sent to sentry per second? Sentry itself will return 429 for any one DSN that sends more than 100 events per second.

Are you capturing an event for each single request your make to your server? The in-memory queue is expected to overflow in this case, it's better than taking over a ton of memory from your server which is obviously already on fire at that point. likely just sending thousands of instances of the same error over and over again.

**Comment by Shtannikov** [comment]
@bruno-garcia We used self-made solution for error logging before Sentry and event per request was fine! At the peak our error logging system processed ~20k rpm (from several hosts of course, from one host ~7k rpm). We expect the same from onprem Sentry.
I have already run synthetic tests with more than 100rps for one DSN and I didn't see 429 in my Nginx logs.

**Comment by bruno-garcia** [comment]
@Shtannikov Sentry.io ingests many thousands of events per seconds across all customers but it's not expected to be receiving 20k events per second from a single client. At that rate the value of the events vs the cost to process/store them is likely gone. It's like running an APM product without any sampling, only works if you have little load or it'll become prohibitly expensive and affect the perf for little benefit.

Also in the event of latency you need to have a back pressure mechanism. In this SDK is the in-memory queue which is optimized for server applications. Storing on disk is another approach taken on the mobile SDKs.

This goes back to [Sentry vs Logging](https://sentry.io/vs/logging/). Sentry isn't trying to replace your logging tool where you dumb raw log entries and index some parts of it. It keeps state in-process and logging integrations keep related log entries (when using Scopes, see ASP.NET Core integration) so in the case of an error, the N previous log messages are all sent together with that event.

**Comment by bruno-garcia** [comment]
I'll close this as dropping events with too high RPM is by design, the back pressure mechanism.

That said, feel free to open a new issue where we could track improving the event submission to have a variable/configurable number of requests in-flight to Sentry to increase the rate one client can submit.
